### ISSUE TITLE
regression on BruteForcePackager with about 15 products - pack() does not complete

### ISSUE THREAD

**Issue by tyrcho** [skjolber/3d-bin-container-packing#447]
Hi,

Thanks for the last fix, I've tested it also on my side.

I've found a new regression. As shown in this branch, this scenario now does not complete in a reasonable time 
https://github.com/skjolber/3d-bin-container-packing/compare/master...tyrcho:fifteen_products_regression

While on 1.2.14 it worked just fine:
https://github.com/skjolber/3d-bin-container-packing/compare/parent-1.2.14...tyrcho:fifteen_products_test?expand=1

I've tried with both FastBruteForcePackager and BruteForcePackager, they behave the same.

**Comment by skjolber** [comment]
Very interesting data, so basically you're using the brute force packager in a different way than I have been expecting. 

![image](https://user-images.githubusercontent.com/1031478/155227152-e4067d93-f09a-462c-9600-f7b67bef796e.png)

It seems you have a lot of space in a container, and thus the brute force packager will not have to run through that many permutations. 

It is as expected that the brute force packager is a bit slower. But not that the `BruteForcePackager` + parallel version is implemented using a recursive algorithm, so you might see stack overflows with that many boxes.

The `FastBruteForcePackager` now runs on 120-130 ms on my machine. 

Anyways, perhaps the `PlainPackager` would be a better alternative than the `LargestAreaFitFirstPackager`. Most of all some packager with "some randomization" would be good enough for `container volume >> sum of box volumes`

**Comment by tyrcho** [comment]
To elaborate a bit on our use case : we want to know if a list of boxes can reasonnably (in a way which a human would find) fit in a container in about 200ms (to get a price for a delivery order).

So we race in parallel BruteForce and LAFF packagers.

Would you suggest a different approach ?

When relying on only one of the 2 packagers, we had found that some trivial cases were rejected (and these are used in my regression tests).

**Comment by skjolber** [comment]
Well the LAFF has some corner cases which really suck, one being a collection of items where a few span the full height. The plain packager is perhaps better. It would be best to just add a few more heuristics, most of which will run fast.

**Comment by skjolber** [comment]
@tyrcho does the v1 packager run faster now? `FastBruteForcePackager` could get the "do not rewind all the way when rotating" optimization.

**Comment by tyrcho** [comment]
@skjolber what do you mean exactly with v1 ? In the branch 1.X ?

**Comment by skjolber** [comment]
@tyrcho Yes, 1.x branch.

**Comment by tyrcho** [comment]
I compared the performance between 1.2.16 and 1.2.19 and it is very similar

**Comment by tyrcho** [comment]
Cool ! I guess this was released in 2.0.11 ? I'll try again with my benchmarks in the next days

**Comment by tyrcho** [comment]
Well, 2.0.11 sounds promising.

FastBruteForce can now pack efficiently this problem:
container [300x350x300] with 32 boxes: 15x182x182 8 and 50x66x231 24
(1.X does not find a solution)

However I found an annoying bug : https://github.com/skjolber/3d-bin-container-packing/pull/522

**Comment by tyrcho** [comment]
This was done with 2 implementations running concurrently to solve each problem
- BruteForce 1.X or FastBruteForce 2.X
- LAFF 1.X with PlainPackager 2.X - this implementation runs over all distinct rotations of the container

I found out that all my tests pass in 2.X when I remove the rotations of the container, if I add a third implementation (LAFF)

I'll run benchmarks with this strategy and keep you posted.

**Comment by skjolber** [comment]
Hmm is that real data secret? I'd like to understand what is happening.

**Comment by tyrcho** [comment]
[Bench.txt](https://github.com/skjolber/3d-bin-container-packing/files/8919260/Bench.txt)

It's completely anonymous, here is the code. Basically : 3 similar containers are defined. Each of the 3 bench tries to pack a set of boxes in the containers.

By "pack", I mean the concurrent code I described above.

I haven't used JMH with maven in a very long while, but if you setup JMH on the project I could share some more examples.
Here seems to be a great tutorial : https://www.baeldung.com/java-microbenchmark-harness

And I use https://jmh.morethan.io/ to get a nice graph and comparison between 2 runs

**Comment by tyrcho** [comment]
OK, even without the rotations, the performance is still up to 90% worse with 2.X :s

**Comment by skjolber** [comment]
Ok, I'd like to figure out why. But I think the result will be that the way the 1.x packager works, it is just simpler and faster. 

Perhaps it is fair to port it to 2.x, even if I thought that would not be necessary.

**Comment by skjolber** [comment]
@tyrcho but that result was for the brute force packager right? The plain / LAFF 2.x should be sufficiently fast..

**Comment by skjolber** [comment]
Also, do you have some kind of rough estimate on cost / benefit of increasing the processing time vs getting a better result?

**Comment by skjolber** [comment]
@tyrcho general note about box sizes and brute force packagers: Grouping items with the same size together is faster.

**Comment by skjolber** [comment]
@tyrcho should this be reopened?

**Comment by tyrcho** [comment]
It works fine ! I finally upgraded in production to 2.1.0 and the metrics seem good
latency is similar

![image](https://user-images.githubusercontent.com/700260/176710088-71ff2cb8-9176-40d5-8121-29c05df3e841.png)

(one the spike due to JVM warmup is passed)

- CPU usage seems also similar (I need more time to assess the impact but it will not be an issue)

We are using both LAFF and FastBruteForce, and I noticed that LAFF could probably still be optimized, it uses much more CPU than in 1.X

![image](https://user-images.githubusercontent.com/700260/176711070-f5e03409-50ba-4926-a29a-49581e0262f5.png)

**Comment by skjolber** [comment]
Nice stats, very interesting. I would say the it is expected that the 2.x is slower than the 1.x, because 2.x is better. 

I think the Plain packager always should be better than the LAFF packager. Probably the Fast LAFF beats the Plain packager on performance, but stacks less well. The 1.x would beat both on speed, but stack even worse.

So I guess you get what you pay for, so far. The brute force algorithm now is quite more "intelligent" than before, as it now skips out-of-scope rotations and permutations.

**Comment by tyrcho** [comment]
Some more insights after the night

![image](https://user-images.githubusercontent.com/700260/176837469-33b6c5ed-e7a3-4f99-baa7-89beb8bf619e.png)

- We are still spending more time than before in LAFF
- With the new version, most of the time seems to be spent sorting stuff
- p95 and p90 latency is much worse now
We are reaching the 200ms timeout much more often

![image](https://user-images.githubusercontent.com/700260/176837776-9d32f109-898a-428f-95b8-01796be9627b.png)

**Comment by skjolber** [comment]
@tyrcho the sorting really does not scale all that well. But the alternative to sorting is scanning through even more points.

**Comment by tyrcho** [comment]
Yesterday I had to rollback to using 1.X in production, too many requests are not respecting the timeout. I think this is mostly related to #533

**Comment by tyrcho** [comment]
I did a bench comparison between 1.X and 2.X (latest)

- 1.X racing BruteForcePackager and LargestAreaFitFirstPackager (all rotations of container)
- 2X racing FastBruteForcePackager, FastLargestAreaFitFirstPackager and LargestAreaFitFirstPackager (no rotations)

all implementations have a deadline of 200ms 


1.X performance is quite good
```
packerBench22Packets             avgt    3   4.355 ±  2.020  ms/op
packerBench33Packets             avgt    3   2.047 ±  4.875  ms/op
packerBench93Packets             avgt    3   5.149 ±  5.682  ms/op
packerBenchHuge                  avgt    3  12.529 ± 19.132  ms/op
packerBenchImpossible100Packets  avgt    3   0.605 ±  0.193  ms/op
```

2.X performance is great for simple problems but bad for large problems (~100 packets)
```

packerBench22Packets             avgt    3    2.812 ±  1.025  ms/op
packerBench33Packets             avgt    3    3.270 ±  1.870  ms/op
packerBench93Packets             avgt    3   22.655 ± 50.195  ms/op
packerBenchHuge	                avgt    3  207.265 ± 11.900  ms/op
packerBenchImpossible100Packets  avgt    3    0.965 ±  0.763  ms/op
``` 

The problem is the large problems are important orders for which we need to find a decent solution.

I've tried different combinations of Packagers in 2.X and for the moment this one gives the best results

**Comment by tyrcho** [comment]
Hmm writing in issues is a bit like talking to a nice 🐥 :D

I changed the implementation to race FastBruteForcePackager, FastLargestAreaFitFirstPackager(rotations)

```
packerBench22Packets             avgt    3   4.963 ±   8.421  ms/op
packerBench33Packets             avgt    3   6.988 ±  51.535  ms/op
packerBench93Packets             avgt    3  55.156 ±  47.389  ms/op
packerBenchHuge                  avgt    3  65.149 ± 734.196  ms/op
packerBenchImpossible100Packets  avgt    3   0.696 ±   0.176  ms/op

```

**Comment by tyrcho** [comment]
after 30 minutes in production with this implementation `race FastBruteForcePackager, FastLargestAreaFitFirstPackager(rotations)` , the latency is similar to 1.X and CPU usage seems fine too
![image](https://user-images.githubusercontent.com/700260/177568026-6015008d-29c0-4427-94d3-c6b22fd4097c.png)

**Comment by tyrcho** [comment]
that would be great ! I'm glad we have (hopefully) succeeded to upgrade to 2.X so we can benefit from such improvements

**Comment by tyrcho** [comment]
I've deployed yesterday a new version based on 2.1.2, the performance and latency seems slightly better 

<img width="1484" alt="image" src="https://user-images.githubusercontent.com/700260/182794873-9799efe0-2620-4f71-a980-6d68eae47b84.png">

More details in a table:
![perf](https://user-images.githubusercontent.com/700260/182796219-9e7ffe42-b818-4d3a-9863-66153e2acacd.png)

**Comment by skjolber** [comment]
@tyrcho thanks for the update, I think if not faster, it is more deterministics this way, building on the previous sort order instead of mixing together and sorting again.
